### ISSUE TITLE
[Question] Jemalloc and (dynamically linked) C/C++ libraries

### ISSUE THREAD

**Issue by Metalhead33** [jemalloc/jemalloc#1321]
[Clone question of one on StackOverflow.](https://stackoverflow.com/questions/52210185/jemalloc-and-c-c-libraries)

If I create libraries in C++ that introduce new classes, and make use of STL operators, and then create new libraries or applications that depend on those libraries, how can I guarantee that the end product will use Jemalloc to allocate memory?

Do I have to globally replace new and delete in all C++ libraries I create and specifically force all STL containers to use _(je)_malloc, or am I good doing this only in the final program that depends on the libraries?

My end goal is that in the final program, all classes and STL containers are allocated via the jemalloc.

**Comment by davidtgoldblatt** [comment]
If your goal is to "really" replace malloc/free/operator new/... for everyone, doing it by linking with main, or LD_PRELOADing, is the way to go.

If you just want to use it in a couple places (e.g. override operator new for one or two classes), it would probably be best to configure jemalloc with the `--with-jemalloc-prefix` option, and call the mangled versions from within your overloads. Note though that jemalloc does not support being `dlclose`d; we'll leak in that case.

**Comment by Metalhead33** [comment]
Yes, I really do want to globally replace malloc, free, and the operators new and delete, for everything, in every library (C and C++ alike) that I have included - I guess in that case, simply dynamically linking in jemalloc with main will suffice without any explicitly operator overloading _(or even writing my own allocator for STL containers)_, right?

**Comment by davidtgoldblatt** [comment]
Yep, that sounds right (well, depending on what exactly you mean by dynamically linking; but so long as our symbols "win" relative to libc's, it should work).

**Comment by YanB25** [comment]
Sorry about replying to this closed issue. I just want to *double check* whether I am correct (to make sure my program works exactly as I expected).

To enable `jemalloc` globally, we need to:
- Dynamically link `jemalloc`, e.g., via `LD_PRELOAD=/xxx/libjemalloc.so`
- Leave `--with-jemalloc-prefix` unspecified (use default value) when jemalloc is compiled, so that the symbols "match" and "win" libsc's.

In other words, if I provided `--with-jemalloc-prefix`, `jemalloc` will not hook the allocation globally as I wish.
